File: app/src/main/java/com/simon/harmonichackernews/data/LinkPreviewFormatUtils.java
package com.simon.harmonichackernews.data;

import androidx.annotation.Nullable;

import java.util.Locale;

final class LinkPreviewFormatUtils {

    private LinkPreviewFormatUtils() {
    }

    static String formatCount(int count, String singular, String plural) {
        if (count == 1) {
            return "1 " + singular;
        }
        return kFormat(count) + " " + plural;
    }

    static String kFormat(int number) {
        if (number < 1000) {
            return String.valueOf(number);
        }

        double rounded = Math.round((double) number / 100) * 100;
        String result = String.format(Locale.US, "%.1fk", rounded / 1000);
        if (result.endsWith(".0k")) {
            return result.substring(0, result.length() - 3) + "k";
        }
        return result;
    }

    @Nullable
    static String shortenUrl(@Nullable String url) {
        if (url == null) {
            return null;
        }

        String shortenedUrl = url;
        if (shortenedUrl.startsWith("https://")) {
            shortenedUrl = shortenedUrl.substring(8);
        } else if (shortenedUrl.startsWith("http://")) {
            shortenedUrl = shortenedUrl.substring(7);
        }

        if (shortenedUrl.startsWith("www.")) {
            shortenedUrl = shortenedUrl.substring(4);
        }

        return shortenedUrl;
    }
}

File: app/src/main/java/com/simon/harmonichackernews/data/LinkPreviewFormatUtils.kt
package com.simon.harmonichackernews.data

import java.util.Locale

internal object LinkPreviewFormatUtils {
    fun formatCount(count: Int, singular: String, plural: String?): String {
        return if (count == 1) "1 $singular" else "${kFormat(count)} $plural"
    }

    fun kFormat(number: Int): String {
        if (number < 1000) {
            return number.toString()
        }

        val rounded = (Math.round(number.toDouble() / 100) * 100).toDouble()
        val result = String.format(Locale.US, "%.1fk", rounded / 1000)
        if (result.endsWith(".0k")) {
            return result.substring(0, result.length - 3) + "k"
        }
        return result
    }

    fun shortenUrl(url: String?): String? {
        url ?: return null

        var shortenedUrl = when {
            url.startsWith("https://") -> url.substring(8)
            url.startsWith("http://") -> url.substring(7)
            else -> url
        }

        shortenedUrl = shortenedUrl.removePrefix("www.")

        return shortenedUrl
    }
}
